### PR TITLE
Shopware Debug plugin

### PR DESCRIPTION
**Pull request by wesolowski**
Hello,

I wrote a plugin for shopware, which makes the [Debug-Plugin](https://github.com/shopware/shopware/tree/5.1/engine/Shopware/Plugins/Default/Core/Debug) available for [Clockwork](https://github.com/itsgoingd/clockwork-chrome) integration. 
All relevant informations can now be displayed in a chrome-tab which provides all features from the FirePHP-version in a more readable overlay. In addition there are some timelines for informations like how long did it take to load an event or to render the site.

For this I need:
- inside the collectors class the attribute 'utils' (protected) to have it available by extending the class
- two "filter"-events in the Bootstrap.php of the Debug-plugin
  - Shopware_Plugins_Core_Debug_Bootstrap_FilterCollectors in the "registerCollectors"-Methode to change the Collector
  - Shopware_Plugins_Core_Debug_Bootstrap_FilterHandlerRegister in the "getHandlers"-Methode to add the ClockworkHandlers

### CODE REVIEW

File: engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php [hunk 1]
```diff
@@ -141,43 +141,55 @@ public function registerCollectors()
...
         if ($this->Config()->get('logEvents')) {
-            $this->pushCollector(new EventCollector($eventManager, $utils));
+            $collectors[] = new EventCollector($eventManager, $utils);
         }
 
-        foreach ($this->collectors as $collector) {
+        $collectors = Enlight()->Events()->filter(
+            'Shopware_Plugins_Core_Debug_Bootstrap_FilterCollectors',
+            $collectors,
+            array(
```
Comment (bcremer) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php]: Please remove `eventManager`, `utils` and `errorHandler` from the event args. If a custom Collector needs them as a dependency just pull them from the DI-Container.

File: engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php [hunk 2]
```diff
@@ -141,43 +141,55 @@ public function registerCollectors()
...
 
         if ($this->Config()->get('logController')) {
-            $this->pushCollector(new ControllerCollector($eventManager, $utils));
+            $collectors[] = new ControllerCollector($eventManager, $utils);
         }
 
         if ($this->Config()->get('logEvents')) {
-            $this->pushCollector(new EventCollector($eventManager, $utils));
+            $collectors[] = new EventCollector($eventManager, $utils);
         }
```
Comment (bcremer) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php:181]: Please use the `collect()` Event type like this:

```
$collection = new ArrayCollection($collectors);
$collection = Enlight()->Events()->collect('Shopware_Plugins_Core_Debug_Bootstrap_Collect_DebugCollector', $collection);
$collectors = $collection->toArray();
```

File: engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php [hunk 3]
```diff
@@ -217,9 +229,16 @@ public function onStartDispatch(\Enlight_Event_EventArgs $args)
      */
     public function getHandlers(\Enlight_Controller_Request_Request $request)
     {
+        $handlerRegister = Enlight()->Events()->filter(
```
Comment (bcremer) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php]: Please use the collect() Event type. See comment above.

**Comment by hlohaus** [comment]
Nice :) :+1:

File: engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php (hunk 2 above)
Comment (wesolowski) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php:181]: I want to change some Collectors.
With the Collect-Event i can only add Collectors, but i cant replace them.
Here is an example how i´d like to use it: [Debug-Subscriber](https://github.com/wesolowski/shopware-clockwork/blob/master/Subscriber/Debug.php#L40)

Comment (bcremer) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php:181]: @wesolowski you are of course right. Please keep the `filter()` event type.

File: engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php [hunk 4]
```diff
@@ -217,9 +225,20 @@ public function onStartDispatch(\Enlight_Event_EventArgs $args)
...
+        $handlerRegister = $this->get('events')->collect(
+            'Shopware_Plugins_Core_Debug_Bootstrap_FilterHandlerRegister',
+            $handlerRegister
+        );
+
         $handlers = array();
-        if ($this->get('monolog.handler.firephp')->acceptsRequest($request)) {
-            $handlers[] = $this->get('monolog.handler.firephp');
+        foreach ($handlerRegister as $handler) {
+            if ($handler instanceof HandlerInterface && $handler->acceptsRequest($request)) {
```
Comment (bcremer) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php]: `acceptsRequest()` is not part of the `HandlerInterface` and should not be checked here.
`acceptsRequest()` is only available on the `monolog.handler.firephp` and should be checked only on that handler.

Comment (wesolowski) [inline engine/Shopware/Plugins/Default/Core/Debug/Bootstrap.php]: You are right, that is not necessary. FirePhpHandler check acceptsRequest in setUp method

**Comment by janbuecker** [comment]
Thanks @wesolowski 

We decided to close this PR because of time constraints and limited overall benefit to the product. If this change is still important to you, please create a new issue on https://issues.shopware.com so people can up-vote this feature.

**Pull request closed without merging**